### PR TITLE
[DEV-3101] Add backend connection to summary and also add unit test

### PR DESCRIPTION
**Pull request by petrpan26**
Modify backend api and add functionality to do summarize for window operator. Added functionality for mock client and also add unit test

### CODE REVIEW

**Comment by aditya-nambiar** [comment]
Is hopping and tumbling functionality a different PR ?

**Comment by aditya-nambiar** [comment]
Also lets hold this PR till black crow is done, else things like the Window struct will break

File: fennel/testing/executor.py [hunk 1]
```diff
@@ -558,7 +597,7 @@ def generate_sessions_for_df(
                     sorted_df[col_name] = sorted_df[col_name].astype(col_type())
                 elif col_type in [int, float, bool, str]:
                     sorted_df[col_name] = sorted_df[col_name].astype(col_type)
-
+            print(sorted_df)
```
Comment (aditya-nambiar) [inline fennel/testing/executor.py:600]: rm print

File: fennel/testing/executor.py [hunk 2]
```diff
@@ -482,56 +482,94 @@ class WindowStruct:
...
-                    "count": self.count,
                 }
 
         def generate_sessions_for_df(
-            df: pd.DataFrame, timestamp_col: str, gap: int, field: str
+            df: pd.DataFrame,
+            timestamp_col: str,
+            gap: int,
+            field: str,
+            summary: Optional[Summary],
```
Comment (aditya-nambiar) [inline fennel/testing/executor.py:500]: Can you add a docstring

File: fennel/testing/executor.py [hunk 3]
```diff
@@ -482,56 +482,94 @@ class WindowStruct:
...
+                            summary_value.append(value)
                         current_window = WindowStruct(row[timestamp_col])
+                        rows = [row]
 
             # Save the last window information
+            # Save the window information
             window_list.append(current_window.to_dict())  # type: ignore
             timestamp_list.append(current_window.end_time)  # type: ignore
+            if summary is not None:
+                value = summary.summarize_func(pd.DataFrame(rows))
```
Comment (aditya-nambiar) [inline fennel/testing/executor.py:540]: You are collecting all the rows and creating the same dataframe ? What is the point of this ?

File: fennel/testing/executor.py [hunk 4]
```diff
@@ -482,56 +482,94 @@ class WindowStruct:
...
-            return pd.DataFrame(
+            df = pd.DataFrame(
                 {field: window_list, timestamp_col: timestamp_list}
             )
+            if summary is not None:
+                try:
+                    df[summary.column_name] = summary_value
+                except Exception as e:
+                    raise Exception(
+                        f"Error in assign node for column `{summary.column_name}` for pipeline "
```
Comment (aditya-nambiar) [inline fennel/testing/executor.py:552]: This error looks incorrect

File: fennel/testing/executor.py [hunk 5]
```diff
@@ -482,56 +482,94 @@ class WindowStruct:
...
+                # Check the schema of the column
+                try:
+                    field = schema_proto.Field(
+                        name=summary.column_name,
+                        dtype=get_datatype(summary.output_type),
+                    )
+                    validate_field_in_df(field, df, self.cur_pipeline_name)
+                except Exception as e:
+                    raise Exception(
+                        f"Error in assign node for column `{summary.column_name}` for pipeline "
```
Comment (aditya-nambiar) [inline fennel/testing/executor.py:564]: Same here

File: fennel/testing/executor.py (hunk 3 above)
Comment (petrpan26) [reply fennel/testing/executor.py:540]: I collect all the rows of one window to summarize

File: fennel/datasets/datasets.py [hunk 6]
```diff
@@ -875,12 +884,28 @@ def dsschema(self):
...
+            values = {}
         return DSSchema(
             keys=keys,
-            values={},  # type: ignore
+            values=values,  # type: ignore
             timestamp=input_schema.timestamp,
         )
 
+    def summarize(self, column: str, result_type: Type, func: Callable):
+        if self.summary is not None:
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:900]: nit: s/1 summary/one summary

**Review by aditya-nambiar — APPROVED** [review]

**Pull request merged**